Panache and Hibernate Reactive use different Vert.x local contexts

DavideD opened the pull request:
Possible fix for https://github.com/quarkusio/quarkus/issues/47314

Draft because it needs [these changes](https://github.com/hibernate/hibernate-reactive/pull/2208) in Hibernate Reactive (next release).
@tsegismont, or @vietj,  Do these changes make sense?

Thanks

DavideD committed:
Upgrade Hibernate Reactive to 2.4.7.Final

tsegismont commented:
@DavideD it seems there are unrelated changes in this PR, correct?

DavideD commented:
Quarkus `main` changes quickly, but this is the important commit:
https://github.com/quarkusio/quarkus/pull/47441/commits/e67c91b9a433279b8a5d03064ee461d40f5e160e

DavideD commented:
Nevermind, I rebased it and not this is the commit with the changes: 3093299438ba898afb579f6da1c14a7cbe2ba343

tsegismont commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java:
```diff
@@ -170,23 +174,30 @@ public static Uni<Mutiny.Session> getSession() {
         return null;
     }
 
+    private static <T> Map<Key<T>, T> contextualDataMap(ContextInternal vertxContext) {
```
Instead of duplicating this method here, we could make it public in Hibernate Reactive and keep the `ContextualDataStorage.CONTEXTUAL_DATA_KEY` private.

By doing so, the implementation details about how the `ConcurrentMap` is created/retrieved remains hidden, yet the consumers can manipulate the data it holds.

DavideD replied:
Yes, I will do it

DavideD committed:
Update access to the Vert.x local context for Panache

In preparation to the upgrade to Vertx. 5, Hibernate Reactive
doesn't use anymore the deprecated API to access the local context.

This commits updates the way Panache access the context so that the
two extensions can comunicate.

Fix #47314
Relats to the Hibernate Reactive issues:
* https://github.com/hibernate/hibernate-reactive/issues/2176
* https://github.com/hibernate/hibernate-reactive/issues/2174
* https://github.com/hibernate/hibernate-reactive/issues/2207

DavideD commented:
@tsegismont [I moved the method to access the map in ContextualDataStorage](https://github.com/hibernate/hibernate-reactive/pull/2217/commits/4535bf308a1722b83e5b9b04d45d3344b1c1507b), and updated this PR. Would this work?

Do you think I also need to udpate [the following code in VertxContextSafety](https://github.com/quarkusio/quarkus/blob/0a319864875100db0ca5900c948da8f5032c20dc/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/context/VertxContextSafetyToggle.java#L68)? I don't think so but I'm not sure.

tsegismont reviewed (approved):
It looks good to me, thank you @DavideD

tsegismont commented:
> Do you think I also need to udpate [the following code in VertxContextSafety](https://github.com/quarkusio/quarkus/blob/0a319864875100db0ca5900c948da8f5032c20dc/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/context/VertxContextSafetyToggle.java#L68)? I don't think so but I'm not sure.

We could avoid to retrieve the Vert.x context twice, is that what you were thinking about?

FroMage reviewed (commented):
Ah, I suppose this will require a similar fix in my Panache 2 branch, due to me adding stateless sessions support in the same area.

DavideD commented:
> We could avoid to retrieve the Vert.x context twice, is that what you were thinking about?

No, I was just wondering if you see something that might need to be updated because of the new API.

cescoffier commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java:
```diff
@@ -50,8 +52,10 @@ public Key<Session> get() {
     // This key is used to indicate that a reactive session should be opened lazily (when needed) in the current vertx context
-    private static final String SESSION_ON_DEMAND_KEY = "hibernate.reactive.panache.sessionOnDemand";
-    private static final String SESSION_ON_DEMAND_OPENED_KEY = "hibernate.reactive.panache.sessionOnDemandOpened";
+    private static final Key<Boolean> SESSION_ON_DEMAND_KEY = new BaseKey<>(Boolean.class,
```
Do not use the "key"-based API for the context local for the time being (and probably for a long time). Quarkus won't support it — we are not even sure if we will support it in Quarkus 4.

DavideD replied:
In Hibernate Reactive we've been  using it for a while... should I change it there too?
Should I only use `String`? What's the issue with using an interface as `Key`?

DavideD replied:
Or by key-based you mean I shouldn't use a Map? What should I use instead?

cescoffier replied:
Quarkus needs to be able to list keys, copy values and manipulate the map - that's how we make sure what is propagated and what is not. Using Key objects makes this very inconvenient and would need to duplicate the whole process. So, switching to that new API means that some feature will not work with Hibernate Reactive.

tsegismont commented:
> @tsegismont so… is context-storage going away? What is it replaced with?

The common contextual data map is not longer accessible from the public API in Vert.x 5. `ContextInternal` still has a `localContextData()` method that gives you this common map, but it is deprecated.

@vietj and I have started to talk about this with @cescoffier and we need to clarify the requirements.

DavideD commented:
If I understand correctly, there's no real issue with this approach except that we will need to change it as soon as something new has been introduced. I can release Hibernate Reactive with this changes for now, right?

tsegismont commented:
> If I understand correctly, there's no real issue with this approach except that we will need to change it as soon as something new has been introduced. I can release Hibernate Reactive with this changes for now, right?

If there are extensions that depend on how Hibernate Reactive stores contextual data, I would wait for @cescoffier to clarify which ones before releasing. Reading the comments from @FroMage , it sounds like Panache is one of them

DavideD commented:
I think Panache is the only one that needs some coordination with Hibernate Reactive.

FroMage commented:
As long as Panache and HR use the same means of storing the session we're fine. We can change HR to match Panache, or change Panache to match HR, it doesn't matter :)

tsegismont commented:
Apparently there's also a difference in how Vert.x and Quarkus create duplicated contexts (Quarkus supports nested context). That could impact the expectations regarding contextual data. So I'd wait for @cescoffier to chime in

markusdlugi commented:
I don't want to disturb your discussion, but we're still waiting for this as fix for #47314, which is currently present in Quarkus 3.21.3 and 3.21.4 (and soon also 3.22). This bug has the potential to cause serious issues in productive applications, and it's likely that this is also affecting other HR Panache users - we were only lucky to spot it due to our good test coverage. So while I understand and appreciate the desire to find the best solution, please keep this in mind. Maybe it might be worth it to separate any major refactoring work into its own issue? Especially since the actual requirements still seem to be unclear as far as I understood your discussion.

FroMage commented:
Good point. Let's get a fix for this issue, and a test ASAP, and we'll see about further improvements later.

DavideD commented:
I'm going to revert the changes I've introduced in Hibernate Reactive and continue to use the deprecated API. We can replace them when there is a suitable solution.
